pkg/icons/icons.go

```diff
@@ -0,0 +1,80 @@
+package icons
+
+import (
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/go-git/go-billy/v5"
+	"github.com/rancher/charts-build-scripts/pkg/filesystem"
+	"github.com/rancher/charts-build-scripts/pkg/path"
+	"github.com/sirupsen/logrus"
+	helmRepo "helm.sh/helm/v3/pkg/repo"
+)
+
+// DownloadIcons download icons to use locally
+func DownloadIcons(rootFs billy.Filesystem) error {
+	absRepositoryHelmIndexFile := filesystem.GetAbsPath(rootFs, path.RepositoryHelmIndexFile)
+	helmIndexFile, err := helmRepo.LoadIndexFile(absRepositoryHelmIndexFile)
+	if err != nil {
+		return fmt.Errorf("encountered error while trying to load existing index file: %s", err)
+	}
+	helmIndexFile.SortEntries()
+	for name := range helmIndexFile.Entries {
+		latestEntryVersion := helmIndexFile.Entries[name][0]
+		latestIconName := name
+		if strings.Contains(name, "-crd") {
+			continue
+		}
+		latestIcon := latestEntryVersion.Icon
+		if latestIcon == "" || strings.HasPrefix(latestIcon, "file://") {
+			logrus.Infof("Skipping icon download for %s", name)
+			continue
+		}
+		p := fmt.Sprintf("assets/logos/%s", latestIconName)
+		logrus.Infof("Downloading icon download for %s", name)
+		ext, err := download(rootFs, latestIcon, p)
+		if err != nil {
+			logrus.Infof("failed to download icon for %s", name)
+			continue
+		}
+		for _, version := range helmIndexFile.Entries[name] {
+			version.Icon = fmt.Sprintf("file://%s%s", p, ext)
+		}
+	}
+	err = helmIndexFile.WriteFile(absRepositoryHelmIndexFile, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("encountered error while trying to write updated Helm index into index.yaml: %s", err)
+	}
+	return nil
+}
+
+func download(rootFs billy.Filesystem, url, path string) (string, error) {
```

Comment:
It is not needed for `rancher/charts`
But for `rancher/partner-charts` if possible and if we are the ones who will implement this same strategy. 
I would consider a for loop with go functions for the download execution. 
120+ rancher partner charts